compiler/query_structure.rs

```diff
@@ -84,67 +170,81 @@ pub fn extract_query_structure_from(
+        );
+        conjuncts.push(QueryStructureConjunct::Block { index: block_id });
+        conjunction.nested_patterns().iter().for_each(|nested| match nested {
+            NestedPattern::Disjunction(disjunction) => {
+                let branches = disjunction
+                    .branch_ids()
+                    .iter()
+                    .zip(disjunction.conjunctions().iter())
+                    .map(|(id, branch)| self.add_block(Some(*id), branch, block_annotations))
+                    .collect::<Vec<_>>();
+                conjuncts.push(QueryStructureConjunct::Or { branches });
+            }
+            NestedPattern::Negation(negation) => {
+                let inner = self.add_block(None, negation.conjunction(), block_annotations);
+                conjuncts.push(QueryStructureConjunct::Not(inner));
+            }
+            NestedPattern::Optional(_) => {
+                unimplemented_feature!(Optionals);
+            }
+        });
+        QueryStructureConjunction { conjunction: conjuncts }
     }
-    branches[branch_id.0 as usize].as_mut().unwrap().extend_from_slice(conjunction.constraints());
-    conjunction.nested_patterns().iter().for_each(|nested| match nested {
-        NestedPattern::Disjunction(disjunction) => {
-            disjunction.branch_ids().iter().zip(disjunction.conjunctions().iter()).for_each(|(id, branch)| {
-                extract_query_structure_from_branch(branches, *id, branch);
-            })
-        }
-        NestedPattern::Negation(_) => {}
-        NestedPattern::Optional(_) => {
-            unimplemented_feature!(Optionals);
-        }
-    })
-}
 
-fn extend_labels_from<'a>(
-    resolved_labels: &mut HashMap<Label, Type>,
-    vertex_annotations: impl Iterator<Item = (&'a Vertex<Variable>, &'a Arc<BTreeSet<answer::Type>>)>,
-) {
-    resolved_labels.extend(vertex_annotations.filter_map(|(vertex, type_)| {
-        match (vertex.as_label(), type_.iter().exactly_one()) {
-            (Some(label), Ok(type_)) => Some((label.clone(), type_.clone())),
-            _ => None,
-        }
-    }));
-}
+    fn add_block_impl(
+        &mut self,
+        existing_branch_id: Option<BranchID>,
+        constraints: &[Constraint<Variable>],
+        annotations: &TypeAnnotations,
+    ) -> QueryStructureBlockID {
+        self.extend_labels_from(annotations);
+        self.extend_function_calls_syntax_from(constraints);
+        let branch_id = if let Some(BranchID(id)) = existing_branch_id {
+            debug_assert!((id as usize) < self.inner.blocks.len());
+            self.inner.blocks[id as usize] = QueryStructureBlock { constraints: Vec::from(constraints) };
+            QueryStructureBlockID(id)
+        } else {
+            self.inner.blocks.push(QueryStructureBlock { constraints: Vec::from(constraints) });
+            QueryStructureBlockID(self.inner.blocks.len() as u16 - 1)
+        };
+        branch_id
+    }
```

Comment:
For blocks which already have a BranchID (i.e. branches of a disjunction & optional blocks), we assign BlockID = BranchID. This avoids having to translate provenance to BlockIDs.